Add stubby resolver daemon service module

baroncharlus opened the pull request:
This change implements [stubby](https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Daemon+-+Stubby), the DNS-over-TLS stub resolver daemon.
The package itself is maintained by @leenaars.

The motivation for this change was the desire to use `stubby`'s
DNS-over-TLS functionality in tandem with unbound, which requires
passing certain configuration parameters. This module implements those
config parameters by exposing them for use in `configuration.nix.`

This is a reopening of PR #38586, which was based on the incorrect branch.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
networking/stubby.nix: implementing systemd service module for stubby

This change implements stubby, the DNS-over-TLS stub resolver daemon.
The motivation for this change was the desire to use stubby's
DNS-over-TLS funcitonality in tandem with unbound, which requires
passing certain configuration parameters. This module implements those
config parameters by exposing them for use in configuration.nix.

Commit:
networking/stubby.nix: merging back module list

re-merging the module list to remove unecessary changes.

baroncharlus commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,205 @@
+        '';
+      };
+
+      enableNetBindService = mkOption {
```
instead of trying to parse the `listenAddresses`, which required the very inefficient `splitString`, I have simply provided the option to disable this capability if the user knows they are going to bind to ports > 1024.

baroncharlus commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,205 @@
+
+      serviceConfig = {
+        AmbientCapabilities = "${optionalString cfg.enableNetBindService "CAP_NET_BIND_SERVICE"}";
+        CapabilitiesBoundingSet = "${optionalString cfg.enableNetBindService "CAP_NET_BIND_SERVICE"}";
```
this part I'm not 100% sure about. obviously, having `AmbientCapabilities=` in the `false` case isn't the prettiest. however, it does seem to achieve the desired effect (for now, anyway).

Infinisil replied:
Eh, I think providing an option for this is a bit too much. Should be fine to just set the values by default, if the user wants to override it he can still do it with something like `systemd.services.stubby.serviceConfig.AmbientCapabilities = mkForce "";`.

Commit:
networking/stubby.nix: removing unecessary capabilities flag

This change removes the unecessary flag for toggling the capabilities
which allows the daemon to bind to low ports.

Commit:
networking/stubby.nix: adding debug level logging bool

Adding the option to turn on debug logging.

wizeman commented:
Also, it would be useful to mention that the `idleTimeout` option is in `ms` units.

BTW, thanks for contributing this service, I am already using it successfully (with `unbound` as a caching layer)!

baroncharlus commented:
@wizeman both good points. very pleased to hear someone found it useful!

Commit:
networking/stubby.nix: clarifying idleTimeout and adding systemd target

Improving docs to note that idleTimeout is expressed in ms. Adding the
nss-lookup `before' target to the systemd service definition.

Infinisil commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,198 @@
+  fallbacks = concatMapStringsSep "\n  " (x: "- ${x}") cfg.fallbackProtocols;
+  listeners = concatMapStringsSep "\n  " (x: "- ${x}") cfg.listenAddresses;
+  # By default, the recursive resolvers maintained by the getdns
+  # project itself are enabled. More information about both getdns's servers, # as well as third party options for upstream resolvers, can be found here:
```
Missed a line break here

Infinisil commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,198 @@
+
+      fallbackProtocols = mkOption {
+        default = [ "GETDNS_TRANSPORT_TLS" ];
+        type = types.listOf types.str;
```
You could use `types.enum [ "GETDNS_TRANSPORT_TLS" "GETDNS_TRANSPORT_TCP" "GETDNS_TRANSPORT_UDP" ]` here

Infinisil commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,198 @@
+      };
+
+      authenticationMode = mkOption {
+        default = "GETDNS_AUTHENTICATION_REQUIRED";
```
You could change this to a boolean option `authenticationRequired` when there's only the values required/none. But only if you're sure that other options won't be added in the future (i suspect that won't be the case).

Infinisil replied:
And if you don't want to make it a boolean, it can be an enum like above instead.

wizeman commented:
BTW, I forgot to mention that I'm getting the following message in the systemd journal:

```Apr 20 15:26:09 wizylap systemd[1]: /nix/store/dh0mk6yh4nhnmh9j3yqimvwcch5h9sp4-unit-stubby.service/stubby.service:15: Unknown lvalue 'CapabilitiesBoundingSet' in section 'Service'```

baroncharlus commented:
@wizeman whups! You're right. that should be `CapabilityBindingSet`, not `Capabilities`. Thanks!

Commit:
networking/stubby.nix: Restrict options with types.enum

This change restricts fallbackProtocol and authenticationMode to accept
only valid options instead of any list or str types (respectively). This
change also fixes typo in the CapabilityBoundingSet systemd setting.

Commit:
networking/stubby.nix: cleaning up documentation

Cleaning up docs, adding literal tags to settings, and removing
whitespace.

Infinisil commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,213 @@
+  listeners = concatMapStringsSep "\n  " (x: "- ${x}") cfg.listenAddresses;
+
+  # By default, the recursive resolvers maintained by the getdns
+  # project itself are enabled. More information about both getdns's servers, # as well as third party options for upstream resolvers, can be found here:
```
I meant the line break in this comment line, there's a `#` in the middle of it.

Commit:
networking/stubby.nix: fixing missing linebreak in comments

Infinisil commented on nixos/modules/services/networking/stubby.nix:
```diff
@@ -0,0 +1,214 @@
+
+  # By default, the recursive resolvers maintained by the getdns
+  # project itself are enabled. More information about both getdns's servers,
+  # # as well as third party options for upstream resolvers, can be found here:
```
Now there's two `#` here..

baroncharlus replied:
oh dear... what are we going to do with me

Commit:
networking/stubby.nix: cleaning errant comments

infinisil reviewed (approved):
Looks perfect to me now, well done! I have also tested this locally and it worked without problems.

baroncharlus commented:
🎉🎉! thanks for your help, @infinisil and @wizeman.

jfrankenau commented:
I've been running this since a few days together with systemd-resolved and it works fine. I guess you'll just have to squash the commits to have this PR merged.
